core/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnTableScan.scala

```diff
@@ -730,10 +769,18 @@ private[sql] final case class ColumnTableScan(
         throw new UnsupportedOperationException(s"unknown type $sqlType")
     }
     if (notNullVar != null) {
-      val nullCode =
-        s"final int $notNullVar = $decoder.notNull($buffer, $batchOrdinal);"
-      if (moveNext.isEmpty) nullCode
-      else s"$nullCode\nif ($notNullVar == 1) $moveNext"
+      if (isWideSchema) {
+        ctx.addMutableState("int", notNullVar, "")
+        val nullCode =
+          s"$notNullVar = $decoder.notNull($buffer, $batchOrdinal);"
```

Comment:
This variable will be used when actual read happens from the decoder. That itself will be in another method. Clubbing them was causing the problem as sometimes in the case of filters actual reads from the decoder are skipped if the filter does not match. This is why we were getting this problem of ArrayIndexOutOfBounds. So separated moving cursor part and reading part to different methods. Common variables used across these two methods need to be made class variable. Another thing that can be done is the return value from the moveNext() methods. That will require a lot of code change, so I have avoided that route. But as we say wide schema needs a re-visit after 0.9.